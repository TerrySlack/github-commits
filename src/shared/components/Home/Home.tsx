import React, { MouseEvent, ChangeEvent } from "react";

import classes from "./home.module.css";

/*
  Selector to pull the owner and route to display
*/
interface Props {
  onSubmit: (e: MouseEvent<HTMLFormElement>) => void;
  onOwnerChange: (e: ChangeEvent<HTMLInputElement>) => void;
  onRepoChange: (e: ChangeEvent<HTMLInputElement>) => void;
}
export const Home = ({ onSubmit, onOwnerChange, onRepoChange }: Props) => {
  return (
    <div data-testid="home" className={classes["home-container"]}>
      <div>
        <div className={classes["home-welcome"]}>
          <h1 data-testid="home-welcome">
            Welcome to the Git Hub commits review app
          </h1>
        </div>
        <div className={classes["home-welcome-form-container"]}>
          <form
            data-testid="home-form"
            onSubmit={onSubmit}
            className={classes["home-welcome-form"]}
          >
            <input
              data-testid="owner"
              type="text"
              className={classes["home-welcome-text-box"]}
              name="owner"
              onChange={onOwnerChange}
              placeholder="Enter the repository owner"
            />
            <input
              data-testid="repo"
              type="text"
              className={classes["home-welcome-text-box"]}
              name="repo"
              onChange={onRepoChange}
              placeholder="Enter the repository name"
            />
            <button
              data-testid="search"
              type="submit"
              className={`data-test-button ${classes["home-welcome-button-common"]} ${classes["home-welcome-button"]}`}
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};
