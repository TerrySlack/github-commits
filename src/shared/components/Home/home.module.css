.home-container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100vh;
  padding: 2em;
}
.home-welcome {
  display: flex;
  justify-content: center;
  color: #5f9ea0;
  padding: 3em;
  width: 100%;
  border-radius: 50px;
  box-shadow: 120px 80px 40px 20px #0ff;
}
.home-welcome-form-container {
  width: 100%;
  padding: 2em;
}
.home-welcome-form {
  display: flex;
  justify-content: center;
  width: 100%;
}

.home-welcome-form > * {
  margin: 0.5em;
}
.home-welcome-button-common {
  padding-top: 0.375rem;
  padding-bottom: 0.375rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  font-size: 1rem;
  font-weight: 400;
  border-radius: 0.375rem;
  text-align: center;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}
.home-welcome-button {
  background-color: #5f9ea0;
  color: #fff;
  cursor: pointer;
  border: solid 2px silver;
}

.home-welcome-disabled {
  color: #fff;
  pointer-events: none;
  background-color: silver;
  border: solid 2px silver;
  opacity: 0.5;
}

.home-welcome-button:hover {
  background-color: #5f9ea0;
  border-color: #fafad2;
  color: #fafad2;
}

.home-welcome-text-box {
  display: block;
  width: 100%;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #212529;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-radius: 0.375rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}
.home-welcome-text-box:focus {
  color: #212529;
  background-color: #fff;
  border-color: #86b7fe;
  outline: 0;
  box-shadow: 0 0 0 0.25rem #a9cacf;
}
.not-found-form {
  height: 35%;
}

.not-found-common {
  color: #fff;
  background-color: #b22222;
}
.not-found-header {
  font-size: large;
  font-weight: bold;
}
.not-found-message {
  color: #fff;
  background-color: #b22222;
  font-size: medium;
  font-weight: 600;
}
.not-found-owner-repo-container {
  display: flex;
  justify-content: center;
}
.not-found-owner-repo-message {
  font-size: medium;
  font-weight: 500;
}
.not-found-owner-repo {
  text-transform: uppercase;
  font-size: medium;
  font-weight: 800;
  padding: 0.5em;
}

.not-found-try-again-container {
  display: flex;
  justify-content: center;
}
.not-found-try-again-container > * {
  margin: 1em;
}
.not-found-try-again-message {
  display: flex;
  justify-content: center;
  align-items: center;
  color: #fff;
  font-size: medium;
  font-weight: 600;
}
