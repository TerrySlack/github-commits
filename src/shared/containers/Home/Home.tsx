import React, {
  MouseEvent,
  useState,
  ChangeEvent,
  useCallback,
  useEffect,
} from "react";

import { useNavigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { debounce } from "Utils/debounce";
import { getRepoCommits, clearCommits } from "Common/Redux";
import { areEqual } from "Utils/equalityChecks";
import { Home } from "Components/Home";

//static selector.
const navigateUrlSelector = ({ common: { navigateUrl } }) => navigateUrl;

const HomeContainer = function () {
  const navigateUrl = useSelector(navigateUrlSelector, areEqual);
  const navigate = useNavigate();

  const dispatch = useDispatch();
  const [owner, setOwner] = useState<string>("");
  const [repo, setRepo] = useState<string>("");

  const onSubmit = (e: MouseEvent<HTMLFormElement>) => {
    e.preventDefault();
    dispatch(getRepoCommits(owner, repo));
  };

  const onOwnerChange = useCallback(
    debounce((e: ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      setOwner(value);
    }),
    []
  );

  const onRepoChange = useCallback(
    debounce((e: ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      setRepo(value);
    }),
    []
  );

  useEffect(() => {
    //I want to reset the state in commits
    dispatch(clearCommits());
  }, []);

  useEffect(() => {
    //The url for navigation is determined in the getRepoCommits thunk.
    //Need this, in order to safely navigate to the /does/not/exist route if the owner/repo does not exist
    if (navigateUrl.length > 0) navigate(navigateUrl);
  }, [navigateUrl]);

  return (
    <Home
      onSubmit={onSubmit}
      onOwnerChange={onOwnerChange}
      onRepoChange={onRepoChange}
    />
  );
};
export { HomeContainer as Home };
