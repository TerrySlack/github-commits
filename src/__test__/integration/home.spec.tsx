import React, { MouseEvent, ChangeEvent, ReactNode } from "react";
import { Provider as ReduxProvider } from "react-redux";
import configureStore from "redux-mock-store";
import {
  render,
  RenderResult,
  fireEvent,
  // waitForElement,
} from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

import { Home as HomeContainer } from "Containers/Home";
import { Home as HomeComponent } from "Components/Home";

//This is the state for the reducer in the project.
const initialState = {
  common: {
    navigateUrl: "",
    owner: "",
    repo: "",
    commits: [],
    page: 1,
    isEndOfCommits: false,
  },
};

interface Props {
  children: ReactNode | ReactNode[];
}

//Add a provider for tests that need it
const Provider = ({ children }: Props) => {
  const mockStore = configureStore();
  const store = mockStore(initialState);
  return <ReduxProvider store={store}>{children}</ReduxProvider>;
};

//Add a router for tests that need it
const Router = ({ children }: Props) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};

let documentBody: RenderResult;

describe("Home Container", () => {
  beforeEach(() => {
    documentBody = render(
      <Router>
        <Provider>
          <HomeContainer />
        </Provider>
      </Router>
    );
  });

  it("matches snapshot", () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});

interface HomeComponentProps {
  isButtonEnabled: boolean;
  onSubmit: (e: MouseEvent<HTMLFormElement>) => void;
  onOwnerChange: (e: ChangeEvent<HTMLInputElement>) => void;
  onRepoChange: (e: ChangeEvent<HTMLInputElement>) => void;
}

const renderHomeComponent = (props: Partial<HomeComponentProps> = {}) => {
  const defaultProps: HomeComponentProps = {
    onSubmit() {
      return;
    },
    onOwnerChange() {
      return;
    },
    onRepoChange() {
      return;
    },
    isButtonEnabled: false,
  };

  return render(<HomeComponent {...defaultProps} {...props} />);
};

describe("Home Component", () => {
  it("button has class data-test-button", async () => {
    const { findByTestId } = renderHomeComponent();
    const searchButton = await findByTestId("search");
    expect(1).toBe(1);
    expect(searchButton.classList.contains("data-test-button")).toBe(true);
  });

  test("should allow entering a repository owner", async () => {
    const onOwnerChange = jest.fn();

    const { findByTestId } = renderHomeComponent({ onOwnerChange });

    const owner = (await findByTestId("owner")) as HTMLInputElement;

    fireEvent.change(owner, { target: { value: "facebook" } });

    expect(owner.value).toBe("facebook");
    expect(onOwnerChange).toHaveBeenCalledTimes(1);
  });

  test("should allow entering a repository name", async () => {
    const onRepoChange = jest.fn();

    const { findByTestId } = renderHomeComponent({ onRepoChange });

    const owner = (await findByTestId("repo")) as HTMLInputElement;

    fireEvent.change(owner, { target: { value: "react" } });

    expect(owner.value).toBe("react");
    expect(onRepoChange).toHaveBeenCalledTimes(1);
  });

  test("should submit the form with facebook and react", async () => {
    // const onSubmit = jest.fn();
    const onSubmit = jest.fn((e) => e.preventDefault());

    const { findByTestId } = renderHomeComponent({
      onSubmit,
    });

    const username = (await findByTestId("owner")) as HTMLInputElement;
    const password = (await findByTestId("repo")) as HTMLInputElement;
    const search = (await findByTestId("search")) as HTMLButtonElement;
    const homeForm = await findByTestId("home-form");

    fireEvent.change(username, { target: { value: "facebook" } });
    fireEvent.change(password, { target: { value: "react" } });
    fireEvent.click(search);

    // expect(onSubmit).toHaveBeenCalledWith("facebook", "react");
    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(homeForm).toHaveFormValues({
      owner: "facebook",
      repo: "react",
    });
  });
  it("matches snapshot", async () => {
    const { findByTestId } = renderHomeComponent();

    const homeForm = await findByTestId("home-form");

    expect(homeForm).toHaveFormValues({
      owner: "",
      repo: "",
    });
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});
