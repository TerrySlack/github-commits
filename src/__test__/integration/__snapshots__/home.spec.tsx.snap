// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component matches snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="home"
    >
      <div>
        <div>
          <h1
            data-testid="home-welcome"
          >
            Welcome to the Git Hub commits review app
          </h1>
        </div>
        <div>
          <form
            data-testid="home-form"
          >
            <input
              data-testid="owner"
              name="owner"
              placeholder="Enter the repository owner"
              type="text"
            />
            <input
              data-testid="repo"
              name="repo"
              placeholder="Enter the repository name"
              type="text"
            />
            <button
              class="data-test-button undefined undefined"
              data-testid="search"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Home Container matches snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="home"
    >
      <div>
        <div>
          <h1
            data-testid="home-welcome"
          >
            Welcome to the Git Hub commits review app
          </h1>
        </div>
        <div>
          <form
            data-testid="home-form"
          >
            <input
              data-testid="owner"
              name="owner"
              placeholder="Enter the repository owner"
              type="text"
            />
            <input
              data-testid="repo"
              name="repo"
              placeholder="Enter the repository name"
              type="text"
            />
            <button
              class="data-test-button undefined undefined"
              data-testid="search"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</body>
`;
